tests/api/test_primer_pair.py

```diff
@@ -556,3 +554,24 @@ def test_primer_pair_compare(
     assert -expected_by_amplicon_false == PrimerPair.compare(
         this=that, that=this, seq_dict=seq_dict, by_amplicon=False
     )
+
+
+def test_calculate_amplicon_span() -> None:
+    left = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr1", 50, 59))
+    right = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr1", 150, 159))
+    assert PrimerPair.calculate_amplicon_span(left, right) == Span("chr1", 50, 159)
+
+    left = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr2", 50, 59))
+    right = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr3", 150, 159))
+    with pytest.raises(ValueError, match="different references"):
+        PrimerPair.calculate_amplicon_span(left, right)
+
+    left = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr1", 150, 159))
+    right = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr1", 50, 59))
+    with pytest.raises(ValueError, match="Left primer does not start before the right primer"):
+        PrimerPair.calculate_amplicon_span(left, right)
+
+    left = Oligo(name="l", bases="AACCGGTTAAACGTT", tm=60, penalty=1, span=Span("chr1", 150, 164))
+    right = Oligo(name="l", bases="AACCGGTTAA", tm=60, penalty=1, span=Span("chr1", 150, 159))
+    with pytest.raises(ValueError, match="Right primer ends before left primer ends"):
+        PrimerPair.calculate_amplicon_span(left, right)
```

Comment:
**note** I usually try to use `pytest.mark.parametrize` to parallelize test execution as much as possible. 

Tests exit after the first failed assertion, so if you bundle test cases in a single test, you run the risk of obscuring later failures. If you parametrize, you hit as many failures as possible in a single test run, and don't find yourself stuck in a loop of addressing one test case, running the test suite, and discovering a previously hidden failure.

I usually have one test parametrized over "good" cases, with each test case paired with its expected output, and one test parametrized over "bad" cases, where each case is expected to raise an exception.

Since here the `span` is the primary variable being manipulated by each case, I'd write something like the following:

```py
@pytest.mark.parametrize(
    "left_primer,right_primer,expected_amplicon",
    [
        (Span("chr1", 50, 59), Span("chr1", 150, 159), Span("chr1", 50, 159)),
    ]
)
def test_calculate_amplicon_span(left_primer: Span, right_primer: Span, expected_amplicon: Span) -> None:
    actual_amplicon: Span = calculate_amplicon_span(left_primer, right_primer)
    assert actual_amplicon == expected_amplicon

@pytest.mark.parametrize(
    "left_primer,right_primer,error_msg"
    [(Span("chr2", 50, 59), Span("chr3", 150, 159), "different references")]
def test_calculate_amplicon_span_raises(left_primer: Span, right_primer: Span, error_msg: str) -> None:
    with pytest.raises(ValueError, match=error_msg):
        calculate_amplicon_span(left_primer, right_primer)
```

Reply:
This is, to me, a stylistic thing, and I come down very strongly on the other side.  I almost never use `pytest.mark.parametrize` because I personally find that it makes tests less maintainable, harder to read etc.  

If the tests aren't as simple as these, then I tend to break them up into multiple test functions.